matchbox_socket/src/webrtc_socket/native.rs

```diff
@@ -82,125 +83,231 @@ impl Signaller for NativeSignaller {
+                trickle_fut,
+                peer_disconnected_rx,
+            ),
+        }
+    }
 
-            // TODO: maybe use some forward trait instead?
-            message = next_peer_messages_out.next().fuse() => {
-                match message {
-                    Some((channel_index, Some((peer, packet)))) => {
-                        let senders = connected_peers.get_mut(&peer)
-                            .unwrap_or_else(|| panic!("couldn't find data channel for peer {peer}"));
-                        let sender = senders.get_mut(channel_index)
-                            .unwrap_or_else(|| panic!("Unexpected data channel index during send: {channel_index}"));
-                        sender.unbounded_send(packet).unwrap();
-                    },
-                    Some((_, None)) | None => {
-                        // Receiver end of outgoing message channel closed,
-                        // which most likely means the socket was dropped.
-                        // There could probably be cleaner ways to handle this,
-                        // but for now, just exit.
-                        debug!("Outgoing message queue closed");
-                        break;
+    async fn peer_loop(peer_uuid: PeerId, handshake_meta: Self::HandshakeMeta) -> PeerId {
+        let (mut to_peer_message_rx, data_channels, mut trickle_fut, mut peer_disconnected) =
+            handshake_meta;
+
+        assert_eq!(
+            data_channels.len(),
+            to_peer_message_rx.len(),
+            "amount of data channels and receivers differ"
+        );
+
+        let mut message_loop_futs: FuturesUnordered<_> = data_channels
+            .iter()
+            .zip(to_peer_message_rx.iter_mut())
+            .map(|(data_channel, rx)| async move {
+                while let Some(message) = rx.next().await {
+                    trace!("sending packet {:?}", message);
+                    let message = message.clone();
+                    let message = Bytes::from(message);
+                    if let Err(e) = data_channel.send(&message).await {
+                        error!("error sending to data channel: {e:?}")
                     }
                 }
-            }
+            })
+            .collect();
 
-            complete => break
+        loop {
+            select! {
+                _ = peer_disconnected.next() => break,
+
+                _ = message_loop_futs.next() => break,
+                // TODO: this means that the signalling is down, should return an
+                // error
+                _ = trickle_fut => continue,
+            }
         }
+
+        peer_uuid
+    }
+}
+
+async fn complete_handshake(
+    trickle: Arc<CandidateTrickle>,
+    connection: &Arc<RTCPeerConnection>,
+    from_peer_rx: UnboundedReceiver<PeerSignal>,
+    mut wait_for_channels: Pin<Box<Fuse<impl Future<Output = ()>>>>,
+) -> Pin<Box<Fuse<Compat<impl Future<Output = Result<(), Box<dyn Error>>>>>>> {
```

Comment:
wow, this one got quite complex

Reply:
Managed to change `Box<dyn Err>` to `webrtc::Error`, but that's about all that could be done

Reply:
Could we at least factor out the complex type?
e.g. `type X = Pin<Box<Fuse<Compat<impl Future<Output = Result<(), Box<dyn Error>>>>>>>`

Reply:
On a second visit I realised `Compat` shouldn't be present in the function signature, so have removed it giving `Pin<Box<Fuse<impl Future<Output = Result<(), webrtc::Error>>>>>`. We could factor it out but that would really just be moving the problem imo

Reply:
Sometimes clippy will consider this a "complex type" and require it to be factored out as a type. If it doesn't make a chirp then I'm good with it.

Reply:
Clippy seems to be happy with this, think it tends to kick up more of a fuss with branching types